server/src/main/java/org/apache/iotdb/db/query/dataset/NonAlignEngineDataSet.java

```diff
@@ -0,0 +1,354 @@
+                  dataSet.cachedBatchData[index] = batchData;
+                  continue;
+                }
+                else
+                  break;
+              }
+              if (dataSet.offsetArray.get(index) == 0) {
+                rowCount++;
+                if (dataSet.limit > 0) {
+                  dataSet.alreadyReturnedRowNumArray.incrementAndGet(index);
+                }
+              } else {
+                dataSet.offsetArray.decrementAndGet(index);
+              }
+            }
+            if (rowCount == 0) {
+              blockingQueue.put(new Pair(null, null));
+              // set the hasRemaining field in reader to false
+              // tell the Consumer not to submit another task for this reader any more
+              reader.setHasRemaining(false);
+              // remove itself from the QueryTaskPoolManager
+              reader.setManagedByQueryManager(false);
+              return;
+            }
+
+            ByteBuffer timeBuffer = ByteBuffer.allocate(timeBAOS.size());
+            timeBuffer.put(timeBAOS.getBuf(), 0, timeBAOS.size());
+            timeBuffer.flip();
+            ByteBuffer valueBuffer = ByteBuffer.allocate(valueBAOS.size());
+            valueBuffer.put(valueBAOS.getBuf(), 0, valueBAOS.size());
+            valueBuffer.flip();
+
+            Pair<ByteBuffer, ByteBuffer> timeValueBAOSPair = new Pair(timeBuffer, valueBuffer);
+
+            blockingQueue.put(timeValueBAOSPair);
+            // if the queue also has free space, just submit another itself
+            if (blockingQueue.remainingCapacity() > 0) {
+              pool.submit(this);
+            }
+            // the queue has no more space
+            // remove itself from the QueryTaskPoolManager
+            else {
+              reader.setManagedByQueryManager(false);
+            }
+            return;
+          }
+          blockingQueue.put(new Pair(null, null));
+          // set the hasRemaining field in reader to false
+          // tell the Consumer not to submit another task for this reader any more
+          reader.setHasRemaining(false);
+          // remove itself from the QueryTaskPoolManager
+          reader.setManagedByQueryManager(false);
+        }
+      } catch (InterruptedException e) {
+        LOGGER.error("Interrupted while putting into the blocking queue: ", e);
+      } catch (IOException e) {
+        LOGGER.error("Something gets wrong while reading from the series reader: ", e);
+      } catch (Exception e) {
+        LOGGER.error("Something gets wrong: ", e);
+      }
+
+    }
+
+  }
+
+
+  private List<ManagedSeriesReader> seriesReaderWithoutValueFilterList;
+
+  // Blocking queue list for each time value buffer pair
+  private BlockingQueue<Pair<ByteBuffer, ByteBuffer>>[] blockingQueueArray;
+
+  private boolean initialized = false;
+
+  private AtomicIntegerArray offsetArray;
+
+  private int limit;
+
+  private AtomicIntegerArray alreadyReturnedRowNumArray;
```

Comment:
Is it necessary to use atomic objects? Will there be concurrent modifications on the same index? I thought all read tasks are working on different indexes.